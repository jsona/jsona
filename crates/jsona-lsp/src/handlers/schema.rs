use crate::{
    lsp_ext::request::{
        AssociatedSchemaParams, AssociatedSchemaResponse, ListSchemasParams, ListSchemasResponse,
        SchemaInfo,
    },
    world::World,
};
use jsona_util::{environment::Environment, schema::associations::AssociationRule};
use lsp_async_stub::{rpc::Error, Context, Params};

#[tracing::instrument(skip_all)]
pub async fn list_schemas<E: Environment>(
    context: Context<World<E>>,
    params: Params<ListSchemasParams>,
) -> Result<ListSchemasResponse, Error> {
    let p = params.required()?;

    let workspaces = context.workspaces.read().await;
    let ws = workspaces.by_document(&p.document_uri);

    let associations = ws.schemas.associations().read();

    Ok(ListSchemasResponse {
        schemas: associations
            .iter()
            .filter(|(rule, _)| !matches!(rule, AssociationRule::Url(..)))
            .map(|(_, s)| SchemaInfo {
                url: s.url.clone(),
                meta: s.meta.clone(),
            })
            .collect(),
    })
}

#[tracing::instrument(skip_all)]
pub async fn associated_schema<E: Environment>(
    context: Context<World<E>>,
    params: Params<AssociatedSchemaParams>,
) -> Result<AssociatedSchemaResponse, Error> {
    let p = params.required()?;

    let workspaces = context.workspaces.read().await;
    let ws = workspaces.by_document(&p.document_uri);

    Ok(AssociatedSchemaResponse {
        schema: ws
            .schemas
            .associations()
            .association_for(&p.document_uri)
            .map(|s| SchemaInfo {
                url: s.url.clone(),
                meta: s.meta.clone(),
            }),
    })
}
