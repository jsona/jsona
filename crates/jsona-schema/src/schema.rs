use either::Either;
use fancy_regex::Regex;
use indexmap::IndexMap;
use jsona::dom::{KeyOrIndex, Keys, Node};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use std::{collections::HashSet, fmt::Display};

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Default)]
pub struct Schema {
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "$ref")]
    pub ref_value: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "$defs")]
    pub defs: Option<IndexMap<String, Schema>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "type")]
    pub schema_type: Option<OneOrMultiTypes>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nullable: Option<bool>,

    #[serde(skip_serializing_if = "Option::is_none", rename = "default")]
    pub default: Option<Value>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub maximum: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub minimum: Option<f64>,
    #[serde(rename = "exclusiveMaximum", skip_serializing_if = "Option::is_none")]
    pub exclusive_maximum: Option<bool>,
    #[serde(rename = "exclusiveMinimum", skip_serializing_if = "Option::is_none")]
    pub exclusive_minimum: Option<bool>,
    #[serde(rename = "multipleOf", skip_serializing_if = "Option::is_none")]
    pub multiple_of: Option<f64>,

    #[serde(rename = "maxLength", skip_serializing_if = "Option::is_none")]
    pub max_length: Option<u32>,
    #[serde(rename = "minLength", skip_serializing_if = "Option::is_none")]
    pub min_length: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pattern: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub items: Option<OneOrMultiSchemas>,
    #[serde(rename = "maxItems", skip_serializing_if = "Option::is_none")]
    pub max_items: Option<u32>,
    #[serde(rename = "minItems", skip_serializing_if = "Option::is_none")]
    pub min_items: Option<u32>,
    #[serde(rename = "uniqueItems", skip_serializing_if = "Option::is_none")]
    pub unique_items: Option<bool>,
    #[serde(rename = "additionalItems", skip_serializing_if = "Option::is_none")]
    pub additional_items: Option<BoolOrSchema>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub contains: Option<Box<Schema>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub properties: Option<IndexMap<String, Schema>>,
    #[serde(rename = "maxProperties", skip_serializing_if = "Option::is_none")]
    pub max_properties: Option<u32>,
    #[serde(rename = "minProperties", skip_serializing_if = "Option::is_none")]
    pub min_properties: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "patternProperties")]
    pub pattern_properties: Option<IndexMap<String, Schema>>,
    #[serde(
        skip_serializing_if = "Option::is_none",
        rename = "additionalProperties"
    )]
    pub additional_properties: Option<BoolOrSchema>,

    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "enum")]
    pub enum_value: Option<Vec<Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "const")]
    pub const_value: Option<Value>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub examples: Option<Vec<Value>>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "readOnly")]
    pub read_only: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "writeOnly")]
    pub write_only: Option<bool>,

    #[serde(rename = "allOf", skip_serializing_if = "Option::is_none")]
    pub all_of: Option<Vec<Schema>>,
    #[serde(rename = "oneOf", skip_serializing_if = "Option::is_none")]
    pub one_of: Option<Vec<Schema>>,
    #[serde(rename = "anyOf", skip_serializing_if = "Option::is_none")]
    pub any_of: Option<Vec<Schema>>,
    #[serde(rename = "not", skip_serializing_if = "Option::is_none")]
    pub not: Option<Vec<Schema>>,
    #[serde(rename = "if", skip_serializing_if = "Option::is_none")]
    pub if_value: Option<Box<Schema>>,
    #[serde(rename = "then", skip_serializing_if = "Option::is_none")]
    pub then_value: Option<Box<Schema>>,
    #[serde(rename = "else", skip_serializing_if = "Option::is_none")]
    pub else_value: Option<Box<Schema>>,

    #[serde(flatten, skip_serializing_if = "Option::is_none")]
    pub unknown: Option<Map<String, Value>>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(transparent)]
pub struct BoolOrSchema {
    #[serde(with = "either::serde_untagged")]
    pub value: Either<bool, Box<Schema>>,
}

impl Default for BoolOrSchema {
    fn default() -> Self {
        Self {
            value: Either::Left(false),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum SchemaType {
    String,
    Number,
    Boolean,
    Null,
    Object,
    Array,
}

impl SchemaType {
    pub fn from_node(node: &Node) -> Option<Self> {
        let schema_type = match &node {
            Node::Null(v) => {
                if v.is_valid() {
                    SchemaType::Null
                } else {
                    return None;
                }
            }
            Node::Bool(_) => SchemaType::Boolean,
            Node::Number(_) => SchemaType::Number,
            Node::String(_) => SchemaType::String,
            Node::Array(_) => SchemaType::Array,
            Node::Object(_) => SchemaType::Object,
        };
        Some(schema_type)
    }
}

impl Display for SchemaType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let type_str = match self {
            SchemaType::String => "string",
            SchemaType::Number => "number",
            SchemaType::Boolean => "boolean",
            SchemaType::Null => "null",
            SchemaType::Object => "object",
            SchemaType::Array => "array",
        };
        f.write_str(type_str)
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(transparent)]
pub struct OneOrMultiTypes {
    #[serde(with = "either::serde_untagged")]
    value: Either<SchemaType, Vec<SchemaType>>,
}

impl OneOrMultiTypes {
    pub fn new(items: impl Iterator<Item = SchemaType>) -> Self {
        let mut items: Vec<SchemaType> = items.collect();
        if items.len() > 1 {
            Self {
                value: Either::Right(items),
            }
        } else {
            Self {
                value: Either::Left(items.remove(0)),
            }
        }
    }
    pub fn contains(&self, target: &SchemaType) -> bool {
        match self.value.as_ref() {
            Either::Left(value) => value == target,
            Either::Right(values) => values.iter().any(|v| v == target),
        }
    }
    pub fn types(&self) -> HashSet<SchemaType> {
        match self.value.as_ref() {
            Either::Left(value) => [value.clone()].into(),
            Either::Right(values) => values.iter().cloned().collect(),
        }
    }
    pub fn len(&self) -> usize {
        match self.value.as_ref() {
            Either::Left(_) => 1,
            Either::Right(values) => values.len(),
        }
    }
}

impl From<SchemaType> for OneOrMultiTypes {
    fn from(schema_type: SchemaType) -> Self {
        Self {
            value: Either::Left(schema_type),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(transparent)]
pub struct OneOrMultiSchemas {
    #[serde(with = "either::serde_untagged")]
    value: Either<Box<Schema>, Vec<Schema>>,
}

impl OneOrMultiSchemas {
    pub fn new(mut items: Vec<Schema>) -> Self {
        if items.len() > 1 {
            Self {
                value: Either::Right(items),
            }
        } else {
            Self {
                value: Either::Left(Box::new(items.remove(0))),
            }
        }
    }
    pub fn as_ref_vec(&self) -> Vec<&Schema> {
        match self.value.as_ref() {
            Either::Left(schema) => vec![schema],
            Either::Right(schemas) => schemas.iter().collect(),
        }
    }
}

impl Schema {
    pub fn pointer(&self, keys: &Keys) -> Vec<&Schema> {
        let mut result = vec![];
        pointer_impl(&mut result, self, self, keys);
        result
    }
    pub fn maybe_object(&self) -> bool {
        self.schema_type
            .as_ref()
            .map(|v| v.contains(&SchemaType::Object))
            .unwrap_or_default()
    }
    pub fn maybe_array(&self) -> bool {
        self.schema_type
            .as_ref()
            .map(|v| v.contains(&SchemaType::Array))
            .unwrap_or_default()
    }
    pub fn maybe_string(&self) -> bool {
        self.schema_type
            .as_ref()
            .map(|v| v.contains(&SchemaType::String))
            .unwrap_or_default()
    }
    pub fn maybe_number(&self) -> bool {
        self.schema_type
            .as_ref()
            .map(|v| v.contains(&SchemaType::Number))
            .unwrap_or_default()
    }
    pub fn maybe_null(&self) -> bool {
        self.schema_type
            .as_ref()
            .map(|v| v.contains(&SchemaType::Null))
            .unwrap_or_default()
    }
    pub fn maybe_boolean(&self) -> bool {
        self.schema_type
            .as_ref()
            .map(|v| v.contains(&SchemaType::Boolean))
            .unwrap_or_default()
    }
    pub fn one_type(&self) -> Option<SchemaType> {
        self.schema_type
            .as_ref()
            .and_then(|v| v.value.as_ref().left())
            .cloned()
    }
    pub fn types(&self) -> HashSet<SchemaType> {
        self.schema_type
            .as_ref()
            .map(|v| v.types())
            .unwrap_or_default()
    }
    pub fn debug_string(&self) -> String {
        serde_json::to_string(&self).unwrap_or_default()
    }
}

fn pointer_impl<'a>(
    result: &mut Vec<&'a Schema>,
    root_schema: &'a Schema,
    local_schema: &'a Schema,
    keys: &Keys,
) {
    let local_schema = match resolve(root_schema, local_schema) {
        Some(v) => v,
        None => return,
    };
    if let Some(schemas) = local_schema
        .one_of
        .as_ref()
        .or(local_schema.any_of.as_ref())
        .or(local_schema.all_of.as_ref())
    {
        for schema in schemas.iter() {
            pointer_impl(result, root_schema, schema, keys);
        }
    } else {
        match keys.shift() {
            None => {
                result.push(local_schema);
            }
            Some((key, keys)) => match key {
                KeyOrIndex::Index(index) => {
                    if let Some(local_schema) = local_schema.items.as_ref() {
                        match local_schema.value.as_ref() {
                            Either::Left(local_schema) => {
                                pointer_impl(result, root_schema, local_schema, &keys)
                            }
                            Either::Right(schemas) => {
                                if let Some(local_schema) = schemas.get(index) {
                                    pointer_impl(result, root_schema, local_schema, &keys)
                                }
                            }
                        }
                    }
                }
                KeyOrIndex::Key(key) => {
                    if let Some(local_schema) = local_schema
                        .properties
                        .as_ref()
                        .and_then(|v| v.get(key.value()))
                    {
                        pointer_impl(result, root_schema, local_schema, &keys)
                    }
                    if let Some(schemas) = local_schema.pattern_properties.as_ref() {
                        for (pat, local_schema) in schemas.iter() {
                            if let Ok(re) = Regex::new(pat) {
                                if let Ok(true) = re.is_match(key.value()) {
                                    pointer_impl(result, root_schema, local_schema, &keys)
                                }
                            }
                        }
                    }
                    if let Some(local_schema) = local_schema
                        .additional_properties
                        .as_ref()
                        .and_then(|v| v.value.as_ref().right())
                    {
                        pointer_impl(result, root_schema, local_schema, &keys)
                    }
                }
                _ => {}
            },
        }
    }
}

fn resolve<'a>(root_schema: &'a Schema, local_schema: &'a Schema) -> Option<&'a Schema> {
    let schema = match local_schema.ref_value.as_ref() {
        Some(ref_value) => {
            if ref_value == "#" {
                root_schema
            } else {
                match root_schema.defs.as_ref().and_then(|defs| {
                    Regex::new(r#"^#/\$defs/(\w+)$"#)
                        .ok()
                        .and_then(|v| v.captures(ref_value).ok().flatten().and_then(|v| v.get(1)))
                        .and_then(|v| defs.get(v.as_str()))
                }) {
                    Some(v) => v,
                    None => return None,
                }
            }
        }
        None => local_schema,
    };
    Some(schema)
}
