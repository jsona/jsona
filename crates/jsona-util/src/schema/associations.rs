use anyhow::{anyhow, bail};
use jsona::dom::{KeyOrIndex, Node};
use parking_lot::{RwLock, RwLockReadGuard};
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;
use tap::Tap;
use tokio::sync::Semaphore;
use url::Url;

use crate::{
    config::Config,
    environment::Environment,
    util::{to_file_url, GlobRule},
};

use super::cache::Cache;

pub const DEFAULT_SCHEMASTORES: &[&str] =
    &["https://cdn.jsdelivr.net/npm/@jsona/schemastore@latest/index.json"];

pub const SCHEMA_KEY: &str = "@jsonaschema";

pub mod priority {
    pub const STORE: usize = 30;
    pub const CONFIG: usize = 50;
    pub const LSP_CONFIG: usize = 60;
    pub const SCHEMA_FIELD: usize = 75;
    pub const MAX: usize = usize::MAX;
}

pub mod source {
    pub const STORE: &str = "store";
    pub const CONFIG: &str = "config";
    pub const LSP_CONFIG: &str = "lsp_config";
    pub const MANUAL: &str = "manual";
    pub const SCHEMA_FIELD: &str = "$schema";
}

#[derive(Clone)]
pub struct SchemaAssociations<E: Environment> {
    concurrent_requests: Arc<Semaphore>,
    associations: Arc<RwLock<Vec<(AssociationRule, SchemaAssociation)>>>,
    env: E,
    cache: Cache<E, Value>,
}

impl<E: Environment> SchemaAssociations<E> {
    pub(crate) fn new(env: E, cache: Cache<E, Value>) -> Self {
        Self {
            concurrent_requests: Arc::new(Semaphore::new(10)),
            cache,
            env,
            associations: Default::default(),
        }
    }
    pub fn add(&self, rule: AssociationRule, assoc: SchemaAssociation) {
        self.associations.write().push((rule, assoc));
    }

    pub fn retain(&self, f: impl Fn(&(AssociationRule, SchemaAssociation)) -> bool) {
        self.associations.write().retain(f);
    }

    pub fn read(&self) -> RwLockReadGuard<'_, Vec<(AssociationRule, SchemaAssociation)>> {
        self.associations.read()
    }

    pub fn clear(&self) {
        self.associations.write().clear();
    }

    pub async fn add_from_schemastore(&self, url: &Url) -> Result<(), anyhow::Error> {
        let schemastore = self.load_schemastore(url).await?;
        for schema in &schemastore.0 {
            match GlobRule::new(&schema.file_match, [] as [&str; 0]) {
                Ok(rule) => {
                    self.associations.write().push((
                        rule.into(),
                        SchemaAssociation {
                            url: schema.url.clone(),
                            meta: json!({
                                "name": schema.name,
                                "description": schema.description,
                                "source": source::STORE,
                                "catalog_url": url,
                            }),
                            priority: priority::STORE,
                        },
                    ));
                }
                Err(error) => {
                    tracing::warn!(
                        %error,
                        schema_name = %schema.name,
                        source = %url,
                        "invalid glob pattern(s)"
                    );
                }
            }
        }
        Ok(())
    }

    /// Adds the schema from a `@jsonaschema` annotation in the root.
    pub fn add_from_document(&self, doc_url: &Url, root: &Node) {
        self.retain(|(rule, assoc)| match rule {
            AssociationRule::Url(u) => {
                !(u == doc_url && assoc.meta["source"] == source::SCHEMA_FIELD)
            }
            _ => true,
        });
        if let Some(url) = root
            .get(&KeyOrIndex::annotation(SCHEMA_KEY))
            .and_then(|v| v.as_string().cloned())
        {
            if let Some(url) = to_file_url(url.value(), self.env.to_file_path(doc_url)) {
                self.associations.write().push((
                    AssociationRule::Url(doc_url.clone()),
                    SchemaAssociation {
                        url,
                        priority: priority::SCHEMA_FIELD,
                        meta: json!({ "source": source::SCHEMA_FIELD }),
                    },
                ));
            }
        }
    }

    pub fn add_from_config(&self, config: &Config) {
        for schema_rule in &config.rules {
            let file_rule = match schema_rule.file_rule.clone() {
                Some(rule) => rule,
                None => continue,
            };
            if let Some(url) = &schema_rule.url {
                self.associations.write().push((
                    file_rule.into(),
                    SchemaAssociation {
                        url: url.clone(),
                        meta: json!({
                            "source": source::CONFIG,
                        }),
                        priority: priority::CONFIG,
                    },
                ));
            }
        }
    }

    pub fn association_for(&self, file: &Url) -> Option<SchemaAssociation> {
        self.associations
            .read()
            .iter()
            .filter_map(|(rule, assoc)| {
                if rule.is_match(file.as_str()) {
                    Some(assoc.clone())
                } else {
                    None
                }
            })
            .max_by_key(|assoc| assoc.priority)
            .tap(|s| {
                if let Some(schema_association) = s {
                    tracing::debug!(
                        schema.url = %schema_association.url,
                        schema.name = schema_association.meta["name"].as_str().unwrap_or(""),
                        schema.source = schema_association.meta["source"].as_str().unwrap_or(""),
                        "found schema association"
                    );
                }
            })
    }

    async fn load_schemastore(&self, index_url: &Url) -> Result<SchemaStore, anyhow::Error> {
        if let Ok(s) = self.cache.load(index_url, false).await {
            return Ok(serde_json::from_value((*s).clone())?);
        }

        let schemastore = match self
            .fetch_external(index_url)
            .await
            .and_then(|v| serde_json::from_slice(&v).map_err(|e| anyhow!("{}", e)))
        {
            Ok(idx) => idx,
            Err(error) => {
                tracing::warn!(?error, "failed to load schemastore");
                if let Ok(s) = self.cache.load(index_url, true).await {
                    return Ok(serde_json::from_value((*s).clone())?);
                }
                return Err(error);
            }
        };

        if self.cache.is_cache_path_set() {
            if let Err(error) = self
                .cache
                .save(
                    index_url.clone(),
                    Arc::new(serde_json::to_value(&schemastore)?),
                )
                .await
            {
                tracing::warn!(%error, "failed to cache schemastore");
            }
        }

        Ok(schemastore)
    }

    async fn fetch_external(&self, file_url: &Url) -> Result<Vec<u8>, anyhow::Error> {
        let _permit = self.concurrent_requests.acquire().await?;
        let data: Vec<u8> = match file_url.scheme() {
            "http" | "https" => self.env.fetch_file(file_url).await?,
            "file" => {
                self.env
                    .read_file(
                        self.env
                            .to_file_path(file_url)
                            .ok_or_else(|| anyhow!("invalid file path"))?
                            .as_ref(),
                    )
                    .await?
            }
            scheme => bail!("the scheme `{scheme}` is not supported"),
        };
        Ok(data)
    }
}

#[derive(Clone)]
pub enum AssociationRule {
    Glob(GlobRule),
    Regex(Regex),
    Url(Url),
}

impl AssociationRule {
    pub fn glob(pattern: &str) -> Result<Self, anyhow::Error> {
        Ok(Self::Glob(GlobRule::new(&[pattern], &[] as &[&str])?))
    }

    pub fn regex(regex: &str) -> Result<Self, anyhow::Error> {
        Ok(Self::Regex(Regex::new(regex)?))
    }
}

impl From<Regex> for AssociationRule {
    fn from(v: Regex) -> Self {
        Self::Regex(v)
    }
}

impl From<GlobRule> for AssociationRule {
    fn from(v: GlobRule) -> Self {
        Self::Glob(v)
    }
}

impl AssociationRule {
    #[must_use]
    pub fn is_match(&self, text: &str) -> bool {
        match self {
            AssociationRule::Glob(g) => g.is_match(text),
            AssociationRule::Regex(r) => r.is_match(text),
            AssociationRule::Url(u) => u.as_str() == text,
        }
    }
}

#[derive(Debug, Clone)]
pub struct SchemaAssociation {
    pub meta: Value,
    pub url: Url,
    pub priority: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(transparent)]
pub struct SchemaStore(Vec<SchemaStoreMeta>);

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaStoreMeta {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub url: Url,
    #[serde(default)]
    pub file_match: Vec<String>,
}
