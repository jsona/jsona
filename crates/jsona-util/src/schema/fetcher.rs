use anyhow::{anyhow, bail};
use arc_swap::ArcSwap;
use std::{path::PathBuf, sync::Arc};
use tokio::sync::Semaphore;
use url::Url;

use crate::environment::Environment;

#[derive(Clone)]
pub struct Fetcher<E: Environment> {
    env: E,
    cache_path: Arc<ArcSwap<Option<PathBuf>>>,
    concurrent_requests: Arc<Semaphore>,
}

impl<E: Environment> Fetcher<E> {
    pub fn new(env: E) -> Self {
        Self {
            env,
            cache_path: Default::default(),
            concurrent_requests: Arc::new(Semaphore::new(10)),
        }
    }

    pub fn set_cache_path(&self, path: Option<PathBuf>) {
        self.cache_path.swap(Arc::new(path));
    }

    #[tracing::instrument(skip_all, fields(%url))]
    pub async fn fetch(&self, url: &Url) -> Result<Vec<u8>, anyhow::Error> {
        let data: Vec<u8> = match url.scheme() {
            "http" | "https" => self.fetch_file(url).await?,
            "file" => self.read_file(url).await?,
            _ => bail!(anyhow!("invalid url `{url}`")),
        };
        Ok(data)
    }

    async fn read_file(&self, url: &Url) -> Result<Vec<u8>, anyhow::Error> {
        self.env
            .read_file(
                self.env
                    .to_file_path(url)
                    .ok_or_else(|| anyhow!("invalid url `{url}`"))?
                    .as_ref(),
            )
            .await
    }

    async fn fetch_file(&self, url: &Url) -> Result<Vec<u8>, anyhow::Error> {
        let _permit = self.concurrent_requests.acquire().await?;
        if let Some(cache_root) = &**self.cache_path.load() {
            let cache_name = format!("{:x}", md5::compute(url.to_string().as_bytes()));
            let cache_path = cache_root.join(&cache_name);
            if let Ok(data) = self.env.read_file(&cache_path).await {
                return Ok(data);
            }
            if let Ok(data) = self.env.fetch_file(url).await {
                if let Err(err) = self.env.write_file(&cache_path, &data).await {
                    tracing::warn!(
                        "failed to write cache file {}, {}",
                        cache_path.display(),
                        err
                    );
                }
                return Ok(data);
            }
        }
        self.env.fetch_file(url).await
    }
}
