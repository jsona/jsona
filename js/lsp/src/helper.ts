
export function createRpc({ write, log }) {
  const rpc = {
    idx: -1,
    waitings: {},
    timeouts: {},
    async readFile(uri: string) {
      return rpc.send({ method: "lsp/readFile", params: { uri } }) as Promise<Uint8Array>;
    },
    async writeFile(uri: string, content: Uint8Array) {
      return rpc.send({ method: "lsp/writeFile", params: { uri, content } }) as Promise<void>;
    },
    send(req) {
      const id = rpc.idx--;
      req.jsonrpc = "2.0";
      req.id = id;
      return new Promise((resolveFn, rejectFn) => {
        const resolve = v => {
          rpc.clean(id);
          return resolveFn(v)
        }
        const reject = v => {
          rpc.clean(id);
          return rejectFn(v)
        }
        rpc.timeouts[id] = setTimeout(() => reject("Operation timeout"), 10000);
        rpc.waitings[id] = { resolve, reject }
        log('lsp2host', req);
        write(req)
      })
    },
    recv(message) {
      if (typeof message.id === "number" && message.id < 0) {
        const wait = rpc.waitings[message.id];
        if (wait) {
          if (message?.error) {
            wait.reject(message.error?.message || "Unknown error")
          } else {
            wait.resolve(message?.result)
          }
        }
        return true
      }
      return false
    },
    clean(id) {
      delete rpc.waitings[id];
      clearTimeout(rpc.timeouts[id]);
      delete rpc.timeouts[id];
    },
  }
  return rpc;
}


export function createLogger({ debug, topics }) {
  const logger = {
    debug,
    topics,
    level() {
      return this.debug ? "debug" : "info";
    },
    log(topic, message) {
      if (!logger.debug) return;
      if (topics.indexOf(topic) === -1 && topics !== "*") return; 
      let printMessage = message;
      if (typeof process !== "undefined") printMessage = JSON.stringify(message);
      if (message?.jsonrpc && message?.method) {
        console.log(topic, message.method, printMessage);
      } else {
        console.log(topic, printMessage);
      }
    }
  };
  return logger;
}
