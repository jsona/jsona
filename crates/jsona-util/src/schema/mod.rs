pub mod associations;
pub mod fetcher;

use anyhow::anyhow;
use jsona::dom::{Keys, Node};
use parking_lot::Mutex;
use std::{path::PathBuf, sync::Arc};
use url::Url;

use self::associations::SchemaAssociations;
use self::fetcher::Fetcher;
use crate::{environment::Environment, HashMap};

pub use jsona_schema_validator::{
    JSONASchemaValidator, JSONASchemaValue, NodeValidationError, Schema,
};

#[derive(Clone)]
pub struct Schemas<E: Environment> {
    env: E,
    associations: SchemaAssociations<E>,
    fetcher: Fetcher<E>,
    validators: Arc<Mutex<HashMap<Url, Arc<JSONASchemaValidator>>>>,
    schemas: Arc<Mutex<HashMap<Url, Arc<JSONASchemaValue>>>>,
}

impl<E: Environment> Schemas<E> {
    pub fn new(env: E) -> Self {
        let fetcher = Fetcher::new(env.clone());
        Self {
            associations: SchemaAssociations::new(env.clone(), fetcher.clone()),
            fetcher,
            env,
            validators: Arc::new(Mutex::new(HashMap::default())),
            schemas: Arc::new(Mutex::new(HashMap::default())),
        }
    }

    /// Get a reference to the schemas's associations.
    pub fn associations(&self) -> &SchemaAssociations<E> {
        &self.associations
    }

    pub fn env(&self) -> &E {
        &self.env
    }

    pub fn set_cache_path(&self, path: Option<PathBuf>) {
        tracing::debug!("set cache path {:?}", path);
        self.fetcher.set_cache_path(path);
    }
}

impl<E: Environment> Schemas<E> {
    #[tracing::instrument(skip_all, fields(%schema_url))]
    pub async fn validate(
        &self,
        schema_url: &Url,
        value: &Node,
    ) -> Result<Vec<NodeValidationError>, anyhow::Error> {
        let validator = match self.get_validator(schema_url) {
            Some(s) => s,
            None => {
                let schema = self
                    .load_schema(schema_url)
                    .await
                    .map_err(|err| anyhow!("failed to load schema {schema_url} {}", err))?;
                self.add_schema(schema_url, schema.clone());
                self.add_validator(schema_url.clone(), &schema)
                    .map_err(|err| anyhow!("load schema {schema_url} throw {}", err))?
            }
        };
        Ok(validator.validate(value))
    }

    pub fn add_schema(&self, schema_url: &Url, schema: Arc<JSONASchemaValue>) {
        drop(self.schemas.lock().insert(schema_url.clone(), schema));
    }

    pub async fn load_schema(
        &self,
        schema_url: &Url,
    ) -> Result<Arc<JSONASchemaValue>, anyhow::Error> {
        if let Some(s) = self.schemas.lock().get(schema_url).cloned() {
            tracing::debug!(%schema_url, "schema was found in cache");
            return Ok(s);
        }

        let schema: Arc<JSONASchemaValue> =
            match self.fetcher.fetch(schema_url).await.and_then(|v| {
                std::str::from_utf8(&v)
                    .map_err(|v| anyhow!("{}", v))
                    .and_then(|v| v.parse().map_err(|err| anyhow!("{}", err)))
            }) {
                Ok(s) => Arc::new(s),
                Err(error) => {
                    tracing::warn!(?error, "failed to fetch remote schema");
                    return Err(error);
                }
            };

        self.schemas
            .lock()
            .insert(schema_url.clone(), schema.clone());

        Ok(schema)
    }

    #[tracing::instrument(skip_all, fields(%schema_url, %path))]
    pub async fn schemas_at_path(
        &self,
        schema_url: &Url,
        path: &Keys,
    ) -> Result<Vec<Schema>, anyhow::Error> {
        let schema = self.load_schema(schema_url).await?;
        let schemas = schema.pointer(path).into_iter().cloned().collect();
        Ok(schemas)
    }

    fn get_validator(&self, schema_url: &Url) -> Option<Arc<JSONASchemaValidator>> {
        self.validators.lock().get(schema_url).cloned()
    }

    fn add_validator(
        &self,
        schema_url: Url,
        schema: &JSONASchemaValue,
    ) -> Result<Arc<JSONASchemaValidator>, anyhow::Error> {
        let v = Arc::new(JSONASchemaValidator::new(schema)?);
        self.validators.lock().insert(schema_url, v.clone());
        Ok(v)
    }
}
