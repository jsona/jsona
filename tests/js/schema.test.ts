import { parse } from "@jsona/schema";
import { readFixtureSync } from "./helper";

const FILES = [
  "schema.jsona"
];

FILES.forEach(name => {
  test(`parse ${name}`, () => {
    const content = readFixtureSync(name)
    expect(parse(content)).toMatchSnapshot();
  })
})
