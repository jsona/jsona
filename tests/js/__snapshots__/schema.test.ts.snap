// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse schema.jsona 1`] = `
Object {
  "errors": null,
  "value": Object {
    "$defs": Object {
      "Object": Object {
        "properties": Object {
          "k1": Object {
            "type": "string",
          },
          "k2": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "properties": Object {
      "_": Object {
        "properties": Object {
          "value": Object {
            "description": "A value",
            "properties": Object {
              "array": Object {
                "oneOf": Array [
                  Object {
                    "type": "integer",
                  },
                  Object {
                    "type": "string",
                  },
                ],
              },
              "bool": Object {
                "default": true,
                "type": "boolean",
              },
              "float": Object {
                "type": "number",
              },
              "integer": Object {
                "maximum": 10,
                "type": "integer",
              },
              "null": Object {},
              "object1": Object {
                "$ref": "#/$defs/Object",
              },
              "object2": Object {
                "$ref": "#/$defs/Object",
              },
              "object3": Object {
                "patternProperties": Object {
                  ".*": Object {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "string": Object {
                "enum": Array [
                  "apple",
                  "orange",
                  "banana",
                ],
                "type": "string",
              },
            },
            "required": Array [
              "float",
            ],
            "type": "object",
          },
        },
        "type": "object",
      },
      "bool": Object {
        "description": "A bool annotation",
        "properties": Object {
          "value": Object {
            "type": "boolean",
          },
        },
        "type": "object",
      },
      "null": Object {
        "description": "A null annotation",
        "properties": Object {
          "value": Object {
            "type": "null",
          },
        },
        "type": "object",
      },
      "object": Object {
        "description": "A object annotation",
        "properties": Object {
          "value": Object {
            "properties": Object {
              "k1": Object {
                "type": "string",
              },
              "k2": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
        },
        "type": "object",
      },
    },
    "type": "object",
  },
}
`;
